### PR TITLE
Allow UnionAll as input to subtypes and include UnionAll in its output

### PR DESCRIPTION
**Pull request by yuyichao**
Replaces https://github.com/JuliaLang/julia/pull/20084 with tests mostly taken from it.

### CODE REVIEW

File: base/reflection.jl [hunk 1]
```diff
@@ -298,22 +298,45 @@ enumerated types (see `@enum`).
...
+            elseif isa(t, Module)
+                t = t::Module
+                in(t, visited) || _subtypes(t, x, sts, visited)
             end
         end
     end
     return sts
 end
-subtypes(m::Module, x::DataType) = sort(collect(_subtypes(m, x)), by=string)
+subtypes(m::Module, x::Union{DataType,UnionAll}) = sort(collect(_subtypes(m, x)), by=string)
```
Comment (pabloferz) [inline base/reflection.jl]: `sort!(collect(...))`?

**Comment by vtjnash** [comment]
Lgtm. Although none of those type asserts should be impactful.

**Comment by yuyichao** [comment]
Removed the `subtypes(Integer)` test since the `examples` test defines an `Integer` subtype and break this test. Not sure why it causes the test to hang though.

**Comment by yuyichao** [comment]
The hang was causing by trying to serialize/deserialize `ModInt` from the example test module from a previous test. Presumably similar to other test system serialization problems.

**Comment by yuyichao** [comment]
Rebased to resolve conflict with #19964, will probably rebased again after https://github.com/JuliaLang/julia/pull/20006 is merged again in order to use `isabstract` and merge after that.

**Comment by yuyichao** [comment]
Rebased on top of #20006.

File: base/reflection.jl [hunk 2]
```diff
@@ -332,22 +332,41 @@ enumerated types (see `@enum`).
...
+            elseif isa(t, Module)
+                t = t::Module
+                in(t, visited) || _subtypes(t, x, sts, visited)
             end
         end
     end
     return sts
 end
-subtypes(m::Module, x::DataType) = x.abstract ? sort!(collect(_subtypes(m, x)), by=string) : DataType[]
+subtypes(m::Module, x::Union{DataType,UnionAll}) = isabstract(x) ? sort!(collect(_subtypes(m, x)), by=string) : Union{DataType,UnionAll}[]
```
Comment (tkelman) [inline base/reflection.jl]: getting to be a long line, either wrap or give the union a shorter alias?

**Pull request merged**